projects/RabbitMQ.Client/client/impl/Connection.cs

```diff
@@ -484,7 +484,7 @@ public void Dispose()
             {
                 if (IsOpen)
                 {
-                    this.AbortAsync().GetAwaiter().GetResult();
+                    throw new InvalidOperationException("CloseAsync/AbortAsync must be called prior to Dispose");
```

Comment:
Throwing exceptions on dispose is generally not expected behaviour. I'd challange this change.

Reply:
What is the best option, then?

I have run into bizarre CI failures that can only be traced to the use of `.GetAwaiter().GetResult()` within `Dispose()`

Reply:
This is one example, from running the entire testsuite on my local Windows 11 workstation. I have no idea how the operations [here](https://github.com/zgabi/rabbitmq-dotnet-client/blob/stringallocations/projects/Test/Integration/TestAsyncConsumer.cs#L598-L620) could have lead to the following error. Notice that, somehow, the test's `DisposeAsync` method (which calls `CloseAsync`) occurred at _the same time_ as a `BasicCancelAsync` operation 🤔 Each operation would have had to acquire the same `SemaphoreSlim` (since the test was using the shared `_channel` instance when the error happened). This is also why I modified the tests to use dedicated `IChannel` instances per-Task:

[rabbitmq-dotnet-client-1233-stacktrace.txt](https://github.com/rabbitmq/rabbitmq-dotnet-client/files/15395296/rabbitmq-dotnet-client-1233-stacktrace.txt)

Reply:
I could look into bringing back a synchronous `Abort` to be used by `Dispose` in another PR.

Reply:
Generally you either do sync or async, which includes on Dispose you call sync methods not asyncs. Otherwise you'll implement IAsyncDisposable and do it there.

On the other hand, this.AbortAsync().GetAwaiter().GetResult(); Should be the way to go. It will block the current thread as a consequence, which might be problematic if
- The threadpool is maxed out, all threads are blocked (See Threadpool starvation)
- the thread has a synchronization context set and it deadlocks itself (Not entirely sure about the details here, rarely happens wihtout an UI AFAIK)


So I don't know exactly based on this why it happens without looking deep into it.

Sidenote: this would be a change I rather see separate from this PR, as I see the intent of enfocing the close as a different breaking change rather than  it being burried in a large PR with a different goal in mind.

Reply:
`GetAwaiter().GetResult()` per-se should not throw exceptions. Exceptions come from task being observed.

The exception in question seems to be coming from [here](https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/929047d5eef58f238e2162ac6a2fddb55d2b985b/projects/RabbitMQ.Client/client/impl/RpcContinuationQueue.cs#L79)

If you are thinking of providing synchronous APIs, they should not be really sync and not sync over async. If it's just for the sake of implementing `IDisposable`, implement just `IAsyncDisposable`.